tool/modules/doip.py

```diff
@@ -0,0 +1,973 @@
+       Returns a list of found service IDs.
+
+    :param arb_id_request: arbitration ID for requests
+    :param arb_id_response: arbitration ID for responses
+    :param timeout: delay between each request sent
+    :param min_id: first service ID to scan
+    :param max_id: last service ID to scan
+    :param print_results: whether progress should be printed to stdout
+    :type arb_id_request: int
+    :type arb_id_response: int
+    :type timeout: float
+    :type min_id: int
+    :type max_id: int
+    :type print_results: bool
+    :return: list of supported service IDs
+    :rtype [int]
+    """
+    found_services = []
+    
+    print("Waiting for Vehicle Identification Announcement\n")
+    print("Power cycle your ECU and wait for a few seconds for the broadcast to be received\n")
+    address, announcement = DoIPClient.await_vehicle_announcement()
+    logical_address = arb_id_request
+    ip, port = address
+
+    print("Discovering Services\n")
+    
+    
+    try:
+    
+        for service_id in range(min_id, max_id + 1):
+                        
+            if print_results:
+                print("\rProbing service 0x{0:02x} ({0}/{1}): found {2}"
+                    .format(service_id, max_id, len(found_services)),
+                    end="")
+            stdout.flush()
+            
+            doip_client = DoIPClient(ip, logical_address, client_logical_address  = arb_id_response )
+        
+            conn = DoIPClientUDSConnector(doip_client)
+            
+            s = struct.pack('<h', service_id) 
+            
+            doip_message = struct.pack("!HH", arb_id_response, arb_id_request) + s[:1] + b'\x00'
+            
+            payload_type = 0x8001
+            
+            try:
+                with Client(conn, request_timeout = timeout) as client:
+                    extended_session(client, session_type = 3)
+                    
+                    doip_client.send_doip(payload_type, doip_message)
+                    response = doip_client.receive_diagnostic(timeout)
+                    
+                    doip_client.close()
+                    
+                    if response[2] == 17 or response is None:
+                        continue
+                    
+                    if response[2] == 18 or response[2] == 19:
+                        response_id = service_id
+                        response_service_id = response[2]
+                        status = response[2]
+                        if response_id != Constants.NR_SI:
+                            request_id = service_id
+                            found_services.append(request_id)
+                        elif status != NegativeResponseCodes.SERVICE_NOT_SUPPORTED:
+                            # Any other response than "service not supported" counts
+                            found_services.append(response_service_id)
+                    
+            except ConfigError:
```

Comment:
Basically this has to do with 2 things:

- I could not disable udsoncan logger, so I was getting big error messages that wasn't helping somehow
- A lot of exceptions are just handling of behavior that was needed to enumerate the unit. for example I was getting the configerror for the first 10-20 services of the 0x00-0xff range, and then it was working properly. I could only fix it and burry the error messages (false positives probably) with the configerror exception.

Basically, with the code you see I managed to get uncluttered results, with meaningful errors, and avoiding the meaningless errors. As I said in my previous module, I am not a developer, so it would be nice if you can propose a more elegant solution :)

Reply:
Seeing as you have a motivation for this pattern and I cannot setup a suitable test environment, I don't really dare to suggest any exact changes in the code for exceptions which I cannot trigger. Let's keep it like this.